flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/serialization/SpanningWrapper.java

```diff
@@ -82,187 +94,171 @@ public SpanningWrapper(String[] tempDirs) {
+		Tuple2<Integer, Integer> positionAndRemaining = readLengthIfNeeded(segment, offset, numBytes);
+		int startPos = positionAndRemaining.f0;
+		int remaining = positionAndRemaining.f1;
+		if (remaining == 0) {
+			return;
 		}
 
-		// copy as much as we need or can for this next spanning record
-		int needed = this.recordLength - this.accumulatedRecordBytes;
-		int toCopy = Math.min(needed, segmentRemaining);
+		int toCopy = min(recordLength - accumulatedRecordBytes, remaining);
+		if (toCopy > 0) {
+			readData(segment, startPos, toCopy);
+		}
+		if (remaining > toCopy) {
+			leftOverData = segment;
+			leftOverStart = startPos + toCopy;
+			leftOverLimit = offset + numBytes;
+		}
+	}
 
-		if (spillingChannel != null) {
-			// spill to file
-			ByteBuffer toWrite = segment.wrap(segmentPosition, toCopy);
-			FileUtils.writeCompletely(this.spillingChannel, toWrite);
+	private void readData(MemorySegment segment, int offset, int length) throws IOException {
+		if (spillingChannel == null) {
+			readIntoBuffer(segment, offset, length);
 		} else {
-			segment.get(segmentPosition, buffer, this.accumulatedRecordBytes, toCopy);
+			readIntoFile(segment, offset, length);
 		}
+	}
 
-		this.accumulatedRecordBytes += toCopy;
+	private void readIntoFile(MemorySegment segment, int offset, int length) throws IOException {
+		writeCompletely(spillingChannel, segment.wrap(offset, length));
+		accumulatedRecordBytes += length;
+		if (hasFullRecord()) {
+			spillingChannel.close();
+			spillFileReader = new DataInputViewStreamWrapper(new BufferedInputStream(new FileInputStream(spillFile), FILE_BUFFER_SIZE));
+		}
+	}
 
-		if (toCopy < segmentRemaining) {
-			// there is more data in the segment
-			this.leftOverData = segment;
-			this.leftOverStart = segmentPosition + toCopy;
-			this.leftOverLimit = numBytes + offset;
+	private void readIntoBuffer(MemorySegment segment, int offset, int length) {
+		segment.get(offset, buffer, accumulatedRecordBytes, length);
+		accumulatedRecordBytes += length;
+		if (hasFullRecord()) {
+			serializationReadBuffer.setBuffer(buffer, 0, recordLength);
 		}
+	}
 
-		if (accumulatedRecordBytes == recordLength) {
-			// we have the full record
-			if (spillingChannel == null) {
-				this.serializationReadBuffer.setBuffer(buffer, 0, recordLength);
-			}
-			else {
-				spillingChannel.close();
+	private Tuple2<Integer, Integer> readLengthIfNeeded(MemorySegment segment, int offset, int numBytes) throws IOException {
+		if (isReadingLength()) {
+			return readLength(segment, offset, numBytes);
+		} else {
+			return Tuple2.of(offset, numBytes);
+		}
+	}
 
-				BufferedInputStream inStream = new BufferedInputStream(new FileInputStream(spillFile), 2 * 1024 * 1024);
-				this.spillFileReader = new DataInputViewStreamWrapper(inStream);
-			}
+	private Tuple2<Integer, Integer> readLength(MemorySegment segment, int segmentPosition, int segmentRemaining) throws IOException {
+		int numBytes = min(lengthBuffer.remaining(), segmentRemaining);
```

Comment:
nit: `numBytes` has a very different meaning on the call stack in `addNextChunkFromMemorySegment()` method, so naming (collision?) is a bit unfortunate.